WD-Firmware/comrv/comrv.c

```diff
@@ -18,34 +18,853 @@
+      /* convert each candidate to an actual value in pEvictCandidatesList */
+      while (uiCandidates)
+      {
+         /* get the lsb that is set */
+         pEvictCandidatesList[ucIndex] = uiCandidates & (-uiCandidates);
+         /* subtract the lsb that is set */
+         uiCandidates -= pEvictCandidatesList[ucIndex]--;
+         pEvictCandidatesList[ucIndex++] += ucEntryIndex*D_COMRV_DWORD_IN_BITS;
+      }
+   }
+
+   /* set the total size of eviction candidates in the last entry */
+   pEvictCandidatesList[ucNumberOfCandidates] = ucAccumulatedSize;
+
+   return ucNumberOfCandidates;
+}
+
+/**
+* search if a specific token is already loaded to the cache
+*
+* @param token - the token to search for
+* @return if the token is loaded the return value is set to the loaded address
+*         otherwise NULL
+*/
+static void* comrvSearchForLoadedOverlayGroup(comrvOverlayToken_t unToken)
+{
+   u08_t ucEntryIndex;
+
+   /* loop all entries */
+   for (ucEntryIndex = 0 ; ucEntryIndex < D_COMRV_NUM_OF_CACHE_ENTRIES ; ucEntryIndex++)
+   {
+      /* if token already loaded */
+      if (g_stComrvCB.stOverlayCache[ucEntryIndex].unToken.stFields.overlayGroupID == unToken.stFields.overlayGroupID)
+      {
+         /* update eviction parameters */
+         comrvUpdateCacheEvectionParams(ucEntryIndex);
+         /* return the actual function location within the loaded overlay group */
+         return g_stComrvCB.stOverlayCache[ucEntryIndex].pFixedEntryAddress;
+      }
+   }
+   /* overlay group not loaded */
+   return 0;
+}
+
+/**
+* Update a given comrv cache entry was accessed
+*
+* @param entryIndex - the comrv cache entry being accessed
+*
+* @return none
+*/
+static void comrvUpdateCacheEvectionParams(u08_t ucEntryIndex)
+{
+   comrvCacheEntry_t *pCacheEntry;
+
+#ifdef D_COMRV_EVICTION_LRU
+
+   /* there is no need to update if ucEntryIndex is already MRU */
+   if (ucEntryIndex !=  g_stComrvCB.ucMruIndex)
+   {
+      pCacheEntry = &g_stComrvCB.stOverlayCache[ucEntryIndex];
+      /* ucEntryIndex is not the ucLruIndex */
+      if (ucEntryIndex !=  g_stComrvCB.ucLruIndex)
+      {
+         /* update previous item's 'next index' */
+         g_stComrvCB.stOverlayCache[pCacheEntry->unLru.stFields.typPrevLruIndex].unLru.stFields.typNextLruIndex =
+               pCacheEntry->unLru.stFields.typNextLruIndex;
+      }
+      else
+      {
+         /* update the global lru index */
+         g_stComrvCB.ucLruIndex = pCacheEntry->unLru.stFields.typNextLruIndex;
+         /* update the lru item with the previous item index */
+         g_stComrvCB.stOverlayCache[g_stComrvCB.ucLruIndex].unLru.stFields.typPrevLruIndex = D_COMRV_LRU_FIRST_ITEM;
+      }
+      /* update ucEntryIndex previous index */
+      pCacheEntry->unLru.stFields.typPrevLruIndex = g_stComrvCB.ucMruIndex;
+      /* update ucEntryIndex next index - last item (MRU)*/
+      pCacheEntry->unLru.stFields.typNextLruIndex = D_COMRV_LRU_LAST_ITEM;
```

Comment:
change to MRU

Reply:
This is OK, I'm using the terms first LRU (LRU item) and last LRU (MRU item)

Reply:
done in a30593c